### ISSUE TITLE
Regression Failure: MB13

### ISSUE THREAD

**Issue by lintool** [castorini/Anserini#409]
I'm getting the following error on MB13 regression (from indexing scratch):

```
2018-08-14 15:45:54,999 INFO  [pool-2-thread-19] index.IndexCollection$IndexerThread (IndexCollection.java:243) - data/statuses.log.2013-02-08-15.gz: 185018 docs added.
2018-08-14 15:56:55,819 WARN  [main] index.IndexCollection (IndexCollection.java:370) - Unexpected difference between number of indexed documents and index maxDoc.
2018-08-14 15:56:55,819 INFO  [main] index.IndexCollection (IndexCollection.java:373) - # Final Counter Values
2018-08-14 15:56:55,820 INFO  [main] index.IndexCollection (IndexCollection.java:374) - indexed:      203,162,923
2018-08-14 15:56:55,821 INFO  [main] index.IndexCollection (IndexCollection.java:375) - empty:            143,890
2018-08-14 15:56:55,821 INFO  [main] index.IndexCollection (IndexCollection.java:376) - unindexed:     55,894,079
2018-08-14 15:56:55,821 INFO  [main] index.IndexCollection (IndexCollection.java:377) - unindexable:            0
2018-08-14 15:56:55,821 INFO  [main] index.IndexCollection (IndexCollection.java:378) - skipped:       88,516,634
2018-08-14 15:56:55,821 INFO  [main] index.IndexCollection (IndexCollection.java:379) - errors:             2,754
2018-08-14 15:56:55,833 INFO  [main] index.IndexCollection (IndexCollection.java:382) - Total 203,145,943 documents indexed in 08:33:45
2018-08-14 15:57:02,736 - regression_test - INFO - ==========Verifying Index==========
2018-08-14 15:57:02,738 - regression_test - INFO - [Index]: lucene-index.mb13.pos+docvectors+rawdocs
2018-08-14 15:57:06,630 - regression_test - INFO - documents:             203145916
Traceback (most recent call last):
  File "src/main/python/run_regression.py", line 203, in <module>
    verify_index(yaml_data, args.index, args.dry_run)
  File "src/main/python/run_regression.py", line 98, in verify_index
    assert value == yaml_data['index_stats'][stat]
AssertionError
```
I'm at this current commit:

```
commit d7764a12c21de01492dd7967cfe347186f068746
Author: Boris Lin <boruilin224@gmail.com>
Date:   Tue Aug 14 10:00:10 2018 +0800

    Patch for Issue #381: Refactor iterator (#399)
    
    Issue #381 - Iterator depending on NoSuchElementException to signal end of iteration.
    This causes exceptions when we try to apply a functional programming style.
```

This is probably the critical line:

```
2018-08-14 15:56:55,819 WARN  [main] index.IndexCollection (IndexCollection.java:370) - Unexpected difference between number of indexed documents and index maxDoc.
```

@borislin it's probably your code above? before, when you hit an error, we bomb out of that segment... now we continue along?

**Comment by lintool** [comment]
Regressions fail here for mb13 also:

```
commit 05df1d68fc6bfa26f291a18fbc34c8c1e13c1dd6
Author: Peilin Yang <yangpeilyn@gmail.com>
Date:   Mon Aug 13 13:23:24 2018 -0700

    fix runtag for TREC submission (#406)
```
I suppose we could do galloping search through the commits to find out which one broke the regression... but does anyone else have any hypotheses?

**Comment by lintool** [comment]
Regressions failed here for mb13 also:

```
commit 6b19d373634d510e8ac5f5b26b1a3d5c321926e4
Author: Peilin Yang <yangpeilyn@gmail.com>
Date:   Thu Aug 9 13:50:44 2018 -0700

    Remove -optimize option for regression test (#393)
```

**Comment by lintool** [comment]
mb13 regressions pass here:

```
commit fb959bbb5c94c376cebc498a6b95600182e1aaee [PASS]
Author: Jimmy Lin <jimmylin@uwaterloo.ca>
Date:   Mon Aug 6 10:25:17 2018 -0400

    Added +rawdocs to name of regression indexes (#389)
    
    #388 added -storeRawDocs, this patch changes the name of the index to reflect it.
```

**Comment by lintool** [comment]
I've isolated the issue to the following patch:

```
commit 6b19d373634d510e8ac5f5b26b1a3d5c321926e4 [FAIL]
Author: Peilin Yang <yangpeilyn@gmail.com>
Date:   Thu Aug 9 13:50:44 2018 -0700

    Remove -optimize option for regression test (#393)

commit 81582f82685118911192e97c35f9d494f19d0fd7 [PASS]
Author: Peilin Yang <yangpeilyn@gmail.com>
Date:   Thu Aug 9 13:20:22 2018 -0700

    fix search runtag arg; make default runtag as Anserini (#392)

commit d4b3272e7f07fa274e5d8ffd50976beb0c08de52 [PASS]
Author: Victor Yang <547106693@qq.com>
Date:   Wed Aug 8 22:33:15 2018 -0400

    Add Regression Test for Car17 (#379)

commit 135d08c7e026c4455ab595f39fbb956ba34afb9f [PASS]
Author: Victor Yang <547106693@qq.com>
Date:   Tue Aug 7 11:13:45 2018 -0400

    Fix zero-length norm bug with RM3 (#376)
    
    Avoids zero-length feedback documents, which causes division by zero when computing term
    weights. Zero-length feedback documents occur (e.g., with CAR17) when a document has only
    terms that accents (which are indexed, but not selected for feedback).
```

In other words, it's @Peilin-Yang 's fault :)

I understand how this happened now - in MB13 there are duplicate documents... I assume optimize cleans up deleted docs, so it yields a different index stats.

**Comment by lintool** [comment]
It appears that the number of docs are non-deterministic! What??

```
$ target/appassembler/bin/IndexUtils -index lucene-index.mb13.pos+docvectors+rawdocs -stats
Index statistics
----------------
documents:             203145916
documents (non-empty): 203143470
unique terms:          -1
total terms:           2099084597
stored fields:
  contents (indexOption: DOCS_AND_FREQS_AND_POSITIONS, hasVectors: true)
  id (indexOption: DOCS, hasVectors: false)
  lang (indexOption: DOCS, hasVectors: false)
  screen_name (indexOption: DOCS, hasVectors: false)
```

**Comment by lintool** [comment]
Fixed by #415
